components/script/dom/readablestream.rs

```diff
@@ -838,6 +898,74 @@ impl Callback for SourceCancelPromiseRejectionHandler {
     }
 }
 
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+struct ForwardPromiseFulfillmentHandler {
+    #[conditional_malloc_size_of]
+    target: Rc<Promise>,
+}
+
+impl Callback for ForwardPromiseFulfillmentHandler {
+    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, _realm: InRealm, can_gc: CanGc) {
+        self.target.resolve(cx, v, can_gc);
+    }
+}
+
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+struct ForwardPromiseRejectionHandler {
+    #[conditional_malloc_size_of]
+    target: Rc<Promise>,
+}
+
+impl Callback for ForwardPromiseRejectionHandler {
+    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, _realm: InRealm, can_gc: CanGc) {
+        self.target.reject(cx, v, can_gc);
+    }
+}
+
+impl js::gc::Rootable for StartCancelAfterAbortHandler {}
+
+#[derive(JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct StartCancelAfterAbortHandler {
+    source: Dom<ReadableStream>,
+    #[ignore_malloc_size_of = "mozjs"]
+    reason: RootedTraceableBox<Heap<JSVal>>,
+    #[conditional_malloc_size_of]
+    result: Rc<Promise>,
+    #[conditional_malloc_size_of]
+    triggered: Rc<Cell<bool>>,
+}
+
+impl Callback for StartCancelAfterAbortHandler {
+    fn callback(&self, cx: SafeJSContext, _v: SafeHandleValue, realm: InRealm, can_gc: CanGc) {
+        if self.triggered.replace(true) {
+            return;
+        }
+
+        let global = GlobalScope::from_safe_context(cx, realm);
+        let source = self.source.as_rooted();
+        rooted!(in(*cx) let mut reason = UndefinedValue());
+        reason.set(self.reason.get());
+        let cancel_promise = if source.is_readable() {
+            source.cancel(cx, &global, reason.handle(), can_gc)
+        } else {
+            Promise::new_resolved(&global, cx, (), can_gc)
```

Comment:
Is it really correct for the handlers to forward an undefined value to whatever `self.result` is in this case?